### ISSUE TITLE
Mismatched Gradient Update with `tf.gather` Operation

### ISSUE THREAD

**Issue by zhujingyang520** [tensorflow/tensorflow#20723]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.7
- **Python version**: 2.7.14 (Anaconda)
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: CUDA 9.0 / cuDNN 7.0
- **GPU model and memory**: NVIDIA GTX 1080Ti
- **Exact command to reproduce**: See below

### Describe the problem
I met the problem when I use tf.gather to slice into a 1D `codebook`  with the specified index, and generate an output Tensor. The loss function is determined based on the generated output Tensor (In the following code, I simply adopt the reduced sum of output Tensor). Then an optimization phase will be applied to fine tune the `codebook` to minimize the loss. The issue happens there is a mismatched updated codebook results when I use built-in `GradientDescentOptimizer` and manually defined one step update results.

### Source code / logs
```python
import numpy as np
import tensorflow as tf

# Generate the synthetic codebook and index for testing
N = 16
shape = (4000, 4000)
np.random.seed(121)
codebook_np = np.random.randn(N)
index_np = np.random.randint(low=0, high=N, size=shape)
lr = 0.001

# Simple computation graph
codebook = tf.get_variable('codebook', shape=[N], dtype=tf.float32, 
                                              initializer=tf.constant_initializer(codebook_np))
index = tf.constant(index_np, name='index', dtype=tf.int64)
output = tf.gather(codebook, index, name='output')
loss = tf.reduce_sum(output, name='loss')
[grad_codebook] = tf.gradients(loss, codebook)
grad_codebook_dense = tf.convert_to_tensor(grad_codebook, name='grad_codebook_dense')
codebook_update = codebook - lr * grad_codebook_dense
opt = tf.train.GradientDescentOptimizer(lr)
train_op = opt.minimize(loss)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

print('codebook [update]')
print(sess.run(codebook_update))
_ = sess.run(train_op)
print('codebook [train_opt]:')
print(sess.run(codebook))
```
I expect to see these 2 updated codebook be same. However, the logs are shown belowed:
```
codebook [update]
[-1000.6481   -999.8909  -1001.59094 -1002.16235  -999.7602   -998.70636
  -998.9787   -999.86896  -999.7241   -998.3734   -997.3994  -1000.765
  -999.6122  -1000.2615   -997.9392  -1002.35077]
codebook [train_opt]:
[-991.77435 -991.03503 -992.6951  -993.2531  -990.9072  -989.8781
 -990.1441  -991.01355 -990.8721  -989.553   -988.60187 -991.8886
 -990.7627  -991.3968  -989.12897 -993.43713]
```
Roughly 1% difference between the updated codebook. The updated code obtained from `train_opt` is always larger than the manually calculated value.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @angersson: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by angersson** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

**Comment by ppwwyyxx** [comment]
This is just a numerical issue with float32. Replacing `codebook` to float64 in the code above will make the gap disappear.
